// @genkit-ai/tools-common/lib/cjs/src/types/reflection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReflectionCancelActionResponseSchema = exports.ReflectionListValuesResponseSchema = exports.ReflectionListActionsResponseSchema = exports.ReflectionEndInputStreamParamsSchema = exports.ReflectionSendInputStreamChunkParamsSchema = exports.ReflectionCancelActionParamsSchema = exports.ReflectionRunActionParamsSchema = exports.ReflectionListValuesParamsSchema = exports.ReflectionConfigureParamsSchema = exports.ReflectionRunActionStateParamsSchema = exports.ReflectionStreamChunkParamsSchema = exports.ReflectionRegisterParamsSchema = exports.ActionMetadataSchema = void 0;
const zod_1 = require("zod");
const action_1 = require("./action");
Object.defineProperty(exports, "ActionMetadataSchema", { enumerable: true, get: function () { return action_1.ActionMetadataSchema; } });
const apis_1 = require("./apis");
exports.ReflectionRegisterParamsSchema = zod_1.z.object({
    id: zod_1.z.string(),
    pid: zod_1.z.number(),
    name: zod_1.z.string().optional(),
    genkitVersion: zod_1.z.string().optional(),
    reflectionApiSpecVersion: zod_1.z.number().optional(),
    envs: zod_1.z.array(zod_1.z.string()).optional(),
});
exports.ReflectionStreamChunkParamsSchema = zod_1.z.object({
    requestId: zod_1.z.string(),
    chunk: zod_1.z.any(),
});
exports.ReflectionRunActionStateParamsSchema = zod_1.z.object({
    requestId: zod_1.z.string(),
    state: zod_1.z
        .object({
        traceId: zod_1.z.string().optional(),
    })
        .optional(),
});
exports.ReflectionConfigureParamsSchema = zod_1.z.object({
    telemetryServerUrl: zod_1.z.string().optional(),
});
exports.ReflectionListValuesParamsSchema = zod_1.z.object({
    type: zod_1.z.string(),
});
exports.ReflectionRunActionParamsSchema = apis_1.RunActionRequestSchema.extend({
    stream: zod_1.z.boolean().optional(),
    streamInput: zod_1.z.boolean().optional(),
});
exports.ReflectionCancelActionParamsSchema = zod_1.z.object({
    traceId: zod_1.z.string(),
});
exports.ReflectionSendInputStreamChunkParamsSchema = zod_1.z.object({
    requestId: zod_1.z.string(),
    chunk: zod_1.z.any(),
});
exports.ReflectionEndInputStreamParamsSchema = zod_1.z.object({
    requestId: zod_1.z.string(),
});
exports.ReflectionListActionsResponseSchema = zod_1.z.object({
    actions: zod_1.z.record(zod_1.z.string(), action_1.ActionMetadataSchema),
});
exports.ReflectionListValuesResponseSchema = zod_1.z.object({
    values: zod_1.z.record(zod_1.z.any()),
});
exports.ReflectionCancelActionResponseSchema = zod_1.z.object({
    message: zod_1.z.string(),
});
