// genkit-cli/dist/utils/manager-utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveTelemetryServer = resolveTelemetryServer;
exports.startManager = startManager;
exports.getDevEnvVars = getDevEnvVars;
exports.startDevProcessManager = startDevProcessManager;
exports.waitForRuntime = waitForRuntime;
exports.runWithManager = runWithManager;
const telemetry_server_1 = require("@genkit-ai/telemetry-server");
const manager_1 = require("@genkit-ai/tools-common/manager");
const utils_1 = require("@genkit-ai/tools-common/utils");
const get_port_1 = __importStar(require("get-port"));
async function resolveTelemetryServer(options) {
    let telemetryServerUrl = process.env.GENKIT_TELEMETRY_SERVER;
    if (!telemetryServerUrl) {
        const telemetryPort = await (0, get_port_1.default)({ port: (0, get_port_1.makeRange)(4033, 4999) });
        telemetryServerUrl = `http://localhost:${telemetryPort}`;
        await (0, telemetry_server_1.startTelemetryServer)({
            port: telemetryPort,
            traceStore: new telemetry_server_1.LocalFileTraceStore({
                storeRoot: options.projectRoot,
                indexRoot: options.projectRoot,
            }),
            logStore: new telemetry_server_1.LocalFileLogStore({
                storeRoot: options.projectRoot,
                indexRoot: options.projectRoot,
            }),
            corsOrigin: options.corsOrigin,
        });
    }
    return telemetryServerUrl;
}
async function startManager(options) {
    const telemetryServerUrl = options.telemetryServerUrl ?? (await resolveTelemetryServer(options));
    const manager = manager_1.RuntimeManager.create({
        telemetryServerUrl,
        manageHealth: options.manageHealth,
        projectRoot: options.projectRoot,
        experimentalReflectionV2: options.experimentalReflectionV2,
        reflectionV2Port: options.reflectionV2Port,
    });
    return manager;
}
async function getDevEnvVars(projectRoot, options) {
    const telemetryServerUrl = await resolveTelemetryServer({
        projectRoot,
        corsOrigin: options?.corsOrigin,
    });
    const disableRealtimeTelemetry = options?.disableRealtimeTelemetry ?? false;
    const experimentalReflectionV2 = options?.experimentalReflectionV2 ?? false;
    let reflectionV2Port;
    const envVars = {
        GENKIT_TELEMETRY_SERVER: telemetryServerUrl,
        GENKIT_ENV: 'dev',
    };
    if (experimentalReflectionV2) {
        reflectionV2Port = await (0, get_port_1.default)({ port: (0, get_port_1.makeRange)(3200, 3400) });
        envVars.GENKIT_REFLECTION_V2_SERVER = `ws://localhost:${reflectionV2Port}`;
    }
    if (!disableRealtimeTelemetry) {
        envVars.GENKIT_ENABLE_REALTIME_TELEMETRY = 'true';
    }
    return { envVars, reflectionV2Port, telemetryServerUrl };
}
async function startDevProcessManager(projectRoot, command, args, options) {
    const { envVars, reflectionV2Port, telemetryServerUrl } = options?.envVars &&
        options?.telemetryServerUrl &&
        (!options?.experimentalReflectionV2 || options?.reflectionV2Port)
        ? {
            envVars: options.envVars,
            reflectionV2Port: options.reflectionV2Port,
            telemetryServerUrl: options.telemetryServerUrl,
        }
        : await getDevEnvVars(projectRoot, options);
    const disableRealtimeTelemetry = options?.disableRealtimeTelemetry ?? false;
    const experimentalReflectionV2 = options?.experimentalReflectionV2 ?? false;
    const processManager = new manager_1.ProcessManager(command, args, envVars);
    const manager = await manager_1.RuntimeManager.create({
        telemetryServerUrl,
        manageHealth: true,
        projectRoot,
        processManager,
        disableRealtimeTelemetry,
        experimentalReflectionV2,
        reflectionV2Port,
    });
    const processPromise = processManager.start({ ...options });
    if (options?.healthCheck) {
        await waitForRuntime(manager, processPromise, options?.timeout);
    }
    return { manager, processPromise };
}
async function waitForRuntime(manager, processPromise, timeoutMs = 30000) {
    let unsubscribe;
    let timeoutId;
    if (manager.listRuntimes().length > 0) {
        return;
    }
    try {
        const runtimeAddedPromise = new Promise((resolve) => {
            unsubscribe = manager.onRuntimeEvent((event) => {
                if (event === manager_1.RuntimeEvent.ADD) {
                    resolve();
                }
            });
            if (manager.listRuntimes().length > 0) {
                resolve();
            }
        });
        const timeoutPromise = new Promise((_, reject) => {
            timeoutId = setTimeout(() => reject(new Error('Timeout waiting for runtime to be ready')), timeoutMs);
        });
        const processExitedPromise = processPromise.then(() => Promise.reject(new Error('Process exited before runtime was ready')), (err) => Promise.reject(err));
        await Promise.race([
            runtimeAddedPromise,
            timeoutPromise,
            processExitedPromise,
        ]);
    }
    finally {
        if (unsubscribe)
            unsubscribe();
        if (timeoutId)
            clearTimeout(timeoutId);
    }
}
async function runWithManager(projectRoot, fn, options) {
    const useEphemeral = options?.runtimeCommand && options.runtimeCommand.length > 0;
    let manager;
    const oldLevel = utils_1.logger.level;
    try {
        if (useEphemeral) {
            const devEnv = await getDevEnvVars(projectRoot, {
                experimentalReflectionV2: true,
            });
            const { envVars, telemetryServerUrl, reflectionV2Port } = devEnv;
            utils_1.logger.level = 'warn';
            const result = await startDevProcessManager(projectRoot, options.runtimeCommand[0], options.runtimeCommand.slice(1), {
                experimentalReflectionV2: true,
                healthCheck: true,
                envVars,
                telemetryServerUrl,
                reflectionV2Port,
                nonInteractive: true,
            });
            manager = result.manager;
        }
        else {
            manager = await startManager({
                projectRoot,
                manageHealth: false,
            });
        }
    }
    catch (e) {
        utils_1.logger.error('Failed to start manager', e);
        process.exit(1);
    }
    finally {
        if (useEphemeral) {
            utils_1.logger.level = oldLevel;
        }
    }
    try {
        await fn(manager);
    }
    catch (err) {
        utils_1.logger.error('Command exited with an Error:');
        const error = err;
        if (typeof error.data === 'object') {
            const errorStatus = error.data;
            const { code, details, message } = errorStatus;
            utils_1.logger.error(`\tCode: ${code}`);
            utils_1.logger.error(`\tMessage: ${message}`);
            if (details?.traceId) {
                utils_1.logger.error(`\tTrace ID: ${details.traceId}\n`);
            }
        }
        else {
            utils_1.logger.error(`\tMessage: ${error.data}\n`);
        }
        utils_1.logger.error('Stack trace:');
        utils_1.logger.error(`${error.stack}`);
    }
    finally {
        if (manager) {
            await manager.stop();
        }
    }
}
