// @genkit-ai/tools-common/lib/esm/src/server/server.js
import * as trpcExpress from '@trpc/server/adapters/express';
import * as bodyParser from 'body-parser';
import * as clc from 'colorette';
import cors from 'cors';
import express from 'express';
import os from 'os';
import path from 'path';
import { writeToolsInfoFile } from '../utils';
import { logger } from '../utils/logger';
import { toolsPackage } from '../utils/package';
import { downloadAndExtractUiAssets } from '../utils/ui-assets';
import { TOOLS_SERVER_ROUTER } from './router';
const MAX_PAYLOAD_SIZE = 30000000;
const UI_ASSETS_GCS_BUCKET = `https://storage.googleapis.com/genkit-assets`;
const UI_ASSETS_ZIP_FILE_NAME = `${toolsPackage.version}.zip`;
const UI_ASSETS_ZIP_GCS_PATH = `${UI_ASSETS_GCS_BUCKET}/${UI_ASSETS_ZIP_FILE_NAME}`;
const UI_ASSETS_ROOT = path.resolve(os.homedir(), '.genkit', 'assets', toolsPackage.version);
const UI_ASSETS_SERVE_PATH = path.resolve(UI_ASSETS_ROOT, 'ui', 'browser');
const API_BASE_PATH = '/api';
class PushableAsyncIterable {
    queue = [];
    resolvers = [];
    closed = false;
    [Symbol.asyncIterator]() {
        return {
            next: () => this.next(),
        };
    }
    next() {
        if (this.queue.length > 0) {
            return Promise.resolve({ value: this.queue.shift(), done: false });
        }
        if (this.closed) {
            return Promise.resolve({ value: undefined, done: true });
        }
        return new Promise((resolve) => this.resolvers.push(resolve));
    }
    push(value) {
        if (this.closed)
            return;
        if (this.resolvers.length > 0) {
            this.resolvers.shift()({ value, done: false });
        }
        else {
            this.queue.push(value);
        }
    }
    close() {
        this.closed = true;
        while (this.resolvers.length > 0) {
            this.resolvers.shift()({ value: undefined, done: true });
        }
    }
}
const activeInputStreams = new Map();
export function startServer(manager, port) {
    let server;
    const app = express();
    app.use(cors({
        origin: '*',
        allowedHeaders: ['Content-Type'],
        exposedHeaders: ['X-Genkit-Trace-Id'],
    }));
    downloadAndExtractUiAssets({
        fileUrl: UI_ASSETS_ZIP_GCS_PATH,
        extractPath: UI_ASSETS_ROOT,
        zipFileName: UI_ASSETS_ZIP_FILE_NAME,
    });
    app.use(express.static(UI_ASSETS_SERVE_PATH));
    app.post('/api/runAction', bodyParser.json({ limit: MAX_PAYLOAD_SIZE }), async (req, res) => {
        res.setHeader('Content-Type', 'application/json');
        res.statusCode = 200;
        if (manager.disableRealtimeTelemetry) {
            res.flushHeaders();
        }
        try {
            const onTraceIdCallback = !manager.disableRealtimeTelemetry
                ? (traceId) => {
                    res.setHeader('X-Genkit-Trace-Id', traceId);
                    res.setHeader('Transfer-Encoding', 'chunked');
                    res.flushHeaders();
                }
                : undefined;
            const result = await manager.runAction(req.body, undefined, onTraceIdCallback);
            res.end(JSON.stringify(result));
        }
        catch (err) {
            const error = err;
            if (!res.headersSent) {
                res.writeHead(500, {
                    'Content-Type': 'application/json',
                });
            }
            res.end(JSON.stringify({ error: error.data }));
        }
    });
    app.post('/api/streamAction', bodyParser.json({ limit: MAX_PAYLOAD_SIZE }), async (req, res) => {
        res.setHeader('Content-Type', 'text/plain');
        res.setHeader('Transfer-Encoding', 'chunked');
        res.statusCode = 200;
        if (manager.disableRealtimeTelemetry) {
            res.flushHeaders();
        }
        let inputStream;
        const bidi = req.query.bidi === 'true';
        if (bidi) {
            inputStream = new PushableAsyncIterable();
        }
        let capturedTraceId;
        try {
            const onTraceIdCallback = (traceId) => {
                capturedTraceId = traceId;
                if (!manager.disableRealtimeTelemetry) {
                    res.setHeader('X-Genkit-Trace-Id', traceId);
                    res.flushHeaders();
                }
                if (bidi && inputStream) {
                    activeInputStreams.set(traceId, inputStream);
                }
            };
            const result = await manager.runAction(req.body, (chunk) => {
                res.write(JSON.stringify(chunk) + '\n');
            }, onTraceIdCallback, inputStream);
            res.write(JSON.stringify(result));
        }
        catch (err) {
            res.write(JSON.stringify({ error: err.data }));
        }
        finally {
            if (capturedTraceId) {
                activeInputStreams.delete(capturedTraceId);
            }
        }
        res.end();
    });
    app.post('/api/sendBidiInput', bodyParser.json({ limit: MAX_PAYLOAD_SIZE }), (req, res) => {
        const { traceId, chunk } = req.body;
        const stream = activeInputStreams.get(traceId);
        if (stream) {
            stream.push(chunk);
            res.status(200).send('OK');
        }
        else {
            res.status(404).send('Stream not found');
        }
    });
    app.post('/api/endBidiInput', bodyParser.json(), (req, res) => {
        const { traceId } = req.body;
        const stream = activeInputStreams.get(traceId);
        if (stream) {
            stream.close();
            res.status(200).send('OK');
        }
        else {
            res.status(404).send('Stream not found');
        }
    });
    app.post('/api/streamTrace', bodyParser.json({ limit: MAX_PAYLOAD_SIZE }), async (req, res) => {
        const { traceId } = req.body;
        if (!traceId) {
            res.status(400).json({ error: 'traceId is required' });
            return;
        }
        res.setHeader('Content-Type', 'text/plain');
        res.setHeader('Transfer-Encoding', 'chunked');
        res.setHeader('X-Genkit-Trace-Id', traceId);
        res.statusCode = 200;
        res.flushHeaders();
        try {
            await manager.streamTrace({ traceId }, (chunk) => {
                res.write(JSON.stringify(chunk) + '\n');
            });
            res.end();
        }
        catch (err) {
            const error = err;
            if (!res.headersSent) {
                res.writeHead(500, {
                    'Content-Type': 'application/json',
                });
            }
            res.write(JSON.stringify({ error: error.data || { message: error.message } }));
            res.end();
        }
    });
    app.get('/api/__health', (_, res) => {
        res.status(200).send('');
    });
    app.post('/api/__quitquitquit', (_, res) => {
        logger.debug('Shutting down tools API');
        res.status(200).send('Server is shutting down');
        server.close(() => {
            process.exit(0);
        });
    });
    app.use(API_BASE_PATH, trpcExpress.createExpressMiddleware({
        router: TOOLS_SERVER_ROUTER(manager),
        maxBodySize: MAX_PAYLOAD_SIZE,
    }));
    app.all('*', (_, res) => {
        res.status(200).sendFile('/', { root: UI_ASSETS_SERVE_PATH });
    });
    const errorHandler = (error, request, response, next) => {
        if (error instanceof Error) {
            logger.error(error.stack);
        }
        return response.status(500).send(error);
    };
    app.use(errorHandler);
    server = app.listen(port, async () => {
        const uiUrl = 'http://localhost:' + port;
        const projectRoot = manager.projectRoot;
        logger.info(`${clc.green(clc.bold('Project root:'))} ${projectRoot}`);
        logger.info(`${clc.green(clc.bold('Genkit Developer UI:'))} ${uiUrl}`);
        await writeToolsInfoFile(uiUrl, projectRoot);
    });
    return new Promise((resolve) => {
        server.once('close', resolve);
    });
}
