// @genkit-ai/tools-common/lib/cjs/src/utils/logger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
exports.forceStderr = forceStderr;
exports.debugToFile = debugToFile;
const clc = __importStar(require("colorette"));
const path = __importStar(require("path"));
const winston = __importStar(require("winston"));
const defaultTransport = new winston.transports.Console();
function forceStderr() {
    exports.logger.remove(defaultTransport);
    exports.logger.add(new winston.transports.Stream({
        stream: process.stderr,
    }));
}
function debugToFile(dirPath) {
    exports.logger.add(new winston.transports.File({
        filename: dirPath
            ? path.join(dirPath, 'genkit-debug.log')
            : 'genkit-debug.log',
        format: winston.format.combine(winston.format.timestamp(), winston.format.json()),
    }));
    exports.logger.level = 'debug';
}
exports.logger = winston.createLogger({
    level: process.env.DEBUG ? 'debug' : 'info',
    format: winston.format.printf((log) => {
        if (log.level === 'info')
            return log.message;
        let levelColor;
        switch (log.level) {
            case 'error':
                levelColor = clc.red;
                break;
            case 'warn':
                levelColor = clc.yellow;
                break;
            default:
                levelColor = (text) => text.toString();
                break;
        }
        const level = log.level.charAt(0).toUpperCase() + log.level.slice(1);
        return `${clc.bold(levelColor(level))}: ${log.message}`;
    }),
    transports: [defaultTransport],
});
