// @genkit-ai/tools-common/lib/esm/src/manager/manager-v2.js
import EventEmitter from 'events';
import getPort, { makeRange } from 'get-port';
import path from 'path';
import { WebSocketServer } from 'ws';
import { RunActionResponseSchema, } from '../types/action';
import { ReflectionCancelActionParamsSchema, ReflectionCancelActionResponseSchema, ReflectionConfigureParamsSchema, ReflectionEndInputStreamParamsSchema, ReflectionListActionsResponseSchema, ReflectionListValuesParamsSchema, ReflectionListValuesResponseSchema, ReflectionRegisterParamsSchema, ReflectionRunActionParamsSchema, ReflectionRunActionStateParamsSchema, ReflectionSendInputStreamChunkParamsSchema, ReflectionStreamChunkParamsSchema, } from '../types/reflection';
import { logger } from '../utils/logger';
import { BaseRuntimeManager } from './manager';
import { GenkitToolsError, RuntimeEvent, } from './types';
export class RuntimeManagerV2 extends BaseRuntimeManager {
    manageHealth;
    _port;
    wss;
    runtimes = new Map();
    get port() {
        return this._port;
    }
    pendingRequests = new Map();
    streamCallbacks = new Map();
    traceIdCallbacks = new Map();
    eventEmitter = new EventEmitter();
    requestIdCounter = 0;
    constructor(telemetryServerUrl, manageHealth, projectRoot, processManager, disableRealtimeTelemetry = false) {
        super(telemetryServerUrl, projectRoot, processManager, disableRealtimeTelemetry);
        this.manageHealth = manageHealth;
    }
    static async create(options) {
        const manager = new RuntimeManagerV2(options.telemetryServerUrl, options.manageHealth ?? true, options.projectRoot, options.processManager, options.disableRealtimeTelemetry);
        await manager.startWebSocketServer(options.reflectionV2Port);
        return manager;
    }
    async startWebSocketServer(port) {
        if (!port) {
            port = await getPort({ port: makeRange(3200, 3400) });
        }
        this.wss = new WebSocketServer({ port });
        this._port = port;
        logger.info(`Starting reflection server: ws://localhost:${port}`);
        this.wss.on('connection', (ws) => {
            ws.on('error', (err) => logger.error(`WebSocket error: ${err}`));
            ws.on('message', (data) => {
                try {
                    const message = JSON.parse(data.toString());
                    this.handleMessage(ws, message);
                }
                catch (error) {
                    logger.error('Failed to parse WebSocket message:', error);
                }
            });
            ws.on('close', () => {
                this.handleDisconnect(ws);
            });
        });
        return { port };
    }
    handleMessage(ws, message) {
        if ('method' in message) {
            this.handleRequest(ws, message);
        }
        else {
            this.handleResponse(message);
        }
    }
    handleRequest(ws, request) {
        switch (request.method) {
            case 'register':
                this.handleRegister(ws, request);
                break;
            case 'streamChunk':
                this.handleStreamChunk(request);
                break;
            case 'runActionState':
                this.handleRunActionState(request);
                break;
            default:
                logger.warn(`Unknown method: ${request.method}`);
        }
    }
    handleRegister(ws, request) {
        const params = ReflectionRegisterParamsSchema.parse(request.params);
        const runtimeInfo = {
            id: params.id,
            pid: params.pid,
            name: params.name,
            genkitVersion: params.genkitVersion,
            reflectionApiSpecVersion: params.reflectionApiSpecVersion,
            reflectionServerUrl: `ws://localhost:${this.port}`,
            timestamp: new Date().toISOString(),
            projectName: path.basename(this.projectRoot),
        };
        this.runtimes.set(runtimeInfo.id, { ws, info: runtimeInfo });
        this.eventEmitter.emit(RuntimeEvent.ADD, runtimeInfo);
        if (request.id) {
            ws.send(JSON.stringify({
                jsonrpc: '2.0',
                result: {
                    telemetryServerUrl: this.telemetryServerUrl,
                },
                id: request.id,
            }));
        }
    }
    handleStreamChunk(notification) {
        const { requestId, chunk } = ReflectionStreamChunkParamsSchema.parse(notification.params);
        const callback = this.streamCallbacks.get(requestId);
        if (callback) {
            callback(chunk);
        }
    }
    handleRunActionState(notification) {
        const { requestId, state } = ReflectionRunActionStateParamsSchema.parse(notification.params);
        const callback = this.traceIdCallbacks.get(requestId);
        if (callback && state?.traceId) {
            callback(state.traceId);
        }
    }
    handleResponse(response) {
        const pending = this.pendingRequests.get(response.id);
        if (pending) {
            if (response.error) {
                const errorData = response.error.data || {};
                const massagedData = {
                    ...errorData,
                    stack: errorData.details?.stack,
                    data: {
                        genkitErrorMessage: errorData.message,
                        genkitErrorDetails: errorData.details,
                    },
                };
                const error = new GenkitToolsError(response.error.message);
                error.data = massagedData;
                pending.reject(error);
            }
            else {
                let result = response.result;
                if (pending.method === 'listActions') {
                    result = ReflectionListActionsResponseSchema.parse(result);
                }
                else if (pending.method === 'listValues') {
                    result = ReflectionListValuesResponseSchema.parse(result);
                }
                else if (pending.method === 'cancelAction') {
                    result = ReflectionCancelActionResponseSchema.parse(result);
                }
                pending.resolve(result);
            }
            this.pendingRequests.delete(response.id);
        }
        else {
            logger.warn(`Received response for unknown request ID ${response.id}`);
        }
    }
    handleDisconnect(ws) {
        for (const [id, runtime] of this.runtimes.entries()) {
            if (runtime.ws === ws) {
                this.runtimes.delete(id);
                this.eventEmitter.emit(RuntimeEvent.REMOVE, runtime.info);
                break;
            }
        }
    }
    async sendRequest(runtimeId, method, params) {
        const runtime = this.runtimes.get(runtimeId);
        if (!runtime) {
            throw new Error(`Runtime ${runtimeId} not found`);
        }
        switch (method) {
            case 'listValues':
                ReflectionListValuesParamsSchema.parse(params);
                break;
            case 'runAction':
                ReflectionRunActionParamsSchema.parse(params);
                break;
            case 'cancelAction':
                ReflectionCancelActionParamsSchema.parse(params);
                break;
        }
        const id = (++this.requestIdCounter).toString();
        const message = {
            jsonrpc: '2.0',
            method,
            params,
            id,
        };
        return new Promise((resolve, reject) => {
            const timeoutId = setTimeout(() => {
                if (this.pendingRequests.has(id)) {
                    this.pendingRequests.delete(id);
                    reject(new Error(`Request ${id} timed out`));
                }
            }, 30000);
            this.pendingRequests.set(id, {
                method,
                resolve: (value) => {
                    clearTimeout(timeoutId);
                    resolve(value);
                },
                reject: (reason) => {
                    clearTimeout(timeoutId);
                    reject(reason);
                },
            });
            runtime.ws.send(JSON.stringify(message));
        });
    }
    sendNotification(runtimeId, method, params) {
        const runtime = this.runtimes.get(runtimeId);
        if (!runtime) {
            logger.warn(`Runtime ${runtimeId} not found, cannot send notification`);
            return;
        }
        switch (method) {
            case 'configure':
                ReflectionConfigureParamsSchema.parse(params);
                break;
            case 'sendInputStreamChunk':
                ReflectionSendInputStreamChunkParamsSchema.parse(params);
                break;
            case 'endInputStream':
                ReflectionEndInputStreamParamsSchema.parse(params);
                break;
        }
        const message = {
            jsonrpc: '2.0',
            method,
            params,
        };
        runtime.ws.send(JSON.stringify(message));
    }
    listRuntimes() {
        return Array.from(this.runtimes.values()).map((r) => r.info);
    }
    getRuntimeById(id) {
        return this.runtimes.get(id)?.info;
    }
    getMostRecentRuntime() {
        const runtimes = this.listRuntimes();
        if (runtimes.length === 0)
            return undefined;
        return runtimes[runtimes.length - 1];
    }
    getMostRecentDevUI() {
        return undefined;
    }
    onRuntimeEvent(listener) {
        const listeners = [];
        Object.values(RuntimeEvent).forEach((event) => {
            const fn = (rt) => listener(event, rt);
            this.eventEmitter.on(event, fn);
            listeners.push({ event, fn });
        });
        return () => {
            listeners.forEach(({ event, fn }) => {
                this.eventEmitter.off(event, fn);
            });
        };
    }
    async listActions(input) {
        const runtimeId = input?.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error(input?.runtimeId
                ? `No runtime found with ID ${input.runtimeId}.`
                : 'No runtimes found. Make sure your app is running using the `start_runtime` MCP tool or the CLI: `genkit start -- ...`. See getting started documentation.');
        }
        const result = {};
        const response = ReflectionListActionsResponseSchema.parse(await this.sendRequest(runtimeId, 'listActions'));
        for (const key of Object.keys(response.actions)) {
            const action = response.actions[key];
            if (!action.key) {
                action.key = key;
            }
            result[key] = action;
        }
        return result;
    }
    async listValues(input) {
        const runtimeId = input?.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error(input?.runtimeId
                ? `No runtime found with ID ${input.runtimeId}.`
                : 'No runtimes found. Make sure your app is running using `genkit start -- ...`. See getting started documentation.');
        }
        const response = await this.sendRequest(runtimeId, 'listValues', {
            type: input.type,
        });
        return response.values;
    }
    async stop() {
        if (this.wss) {
            this.wss.close();
        }
        if (this.processManager) {
            await this.processManager.kill();
        }
    }
    async runAction(input, streamingCallback, onTraceId, inputStream) {
        const runtimeId = input.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error('No runtimes found. Make sure your app is running using the `start_runtime` MCP tool or the CLI: `genkit start -- ...`. See getting started documentation.');
        }
        const runtime = this.runtimes.get(runtimeId);
        if (!runtime) {
            throw new Error(`Runtime ${runtimeId} not found`);
        }
        const id = (++this.requestIdCounter).toString();
        if (streamingCallback) {
            this.streamCallbacks.set(id, streamingCallback);
        }
        if (onTraceId) {
            this.traceIdCallbacks.set(id, onTraceId);
        }
        const message = {
            jsonrpc: '2.0',
            method: 'runAction',
            params: {
                ...input,
                stream: !!streamingCallback,
                streamInput: !!inputStream,
            },
            id,
        };
        const promise = new Promise((resolve, reject) => {
            this.pendingRequests.set(id, { method: 'runAction', resolve, reject });
            runtime.ws.send(JSON.stringify(message));
        })
            .then((result) => {
            return RunActionResponseSchema.parse(result);
        })
            .finally(() => {
            if (streamingCallback) {
                this.streamCallbacks.delete(id);
            }
            if (onTraceId) {
                this.traceIdCallbacks.delete(id);
            }
        });
        if (inputStream) {
            (async () => {
                try {
                    for await (const chunk of inputStream) {
                        this.sendNotification(runtimeId, 'sendInputStreamChunk', {
                            requestId: id,
                            chunk,
                        });
                    }
                    this.sendNotification(runtimeId, 'endInputStream', {
                        requestId: id,
                    });
                }
                catch (e) {
                    logger.error(`Error streaming input: ${e}`);
                }
            })();
        }
        return promise;
    }
    async cancelAction(input) {
        const runtimeId = input.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error('No runtime found');
        }
        return this.sendRequest(runtimeId, 'cancelAction', {
            traceId: input.traceId,
        });
    }
}
