// @genkit-ai/tools-common/lib/esm/src/types/reflection.js
import { z } from 'zod';
import { ActionMetadataSchema } from './action';
import { RunActionRequestSchema } from './apis';
export { ActionMetadataSchema };
export const ReflectionRegisterParamsSchema = z.object({
    id: z.string(),
    pid: z.number(),
    name: z.string().optional(),
    genkitVersion: z.string().optional(),
    reflectionApiSpecVersion: z.number().optional(),
    envs: z.array(z.string()).optional(),
});
export const ReflectionStreamChunkParamsSchema = z.object({
    requestId: z.string(),
    chunk: z.any(),
});
export const ReflectionRunActionStateParamsSchema = z.object({
    requestId: z.string(),
    state: z
        .object({
        traceId: z.string().optional(),
    })
        .optional(),
});
export const ReflectionConfigureParamsSchema = z.object({
    telemetryServerUrl: z.string().optional(),
});
export const ReflectionListValuesParamsSchema = z.object({
    type: z.string(),
});
export const ReflectionRunActionParamsSchema = RunActionRequestSchema.extend({
    stream: z.boolean().optional(),
    streamInput: z.boolean().optional(),
});
export const ReflectionCancelActionParamsSchema = z.object({
    traceId: z.string(),
});
export const ReflectionSendInputStreamChunkParamsSchema = z.object({
    requestId: z.string(),
    chunk: z.any(),
});
export const ReflectionEndInputStreamParamsSchema = z.object({
    requestId: z.string(),
});
export const ReflectionListActionsResponseSchema = z.object({
    actions: z.record(z.string(), ActionMetadataSchema),
});
export const ReflectionListValuesResponseSchema = z.object({
    values: z.record(z.any()),
});
export const ReflectionCancelActionResponseSchema = z.object({
    message: z.string(),
});
