// @genkit-ai/tools-common/lib/cjs/src/manager/manager-v2.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeManagerV2 = void 0;
const events_1 = __importDefault(require("events"));
const get_port_1 = __importStar(require("get-port"));
const path_1 = __importDefault(require("path"));
const ws_1 = require("ws");
const action_1 = require("../types/action");
const reflection_1 = require("../types/reflection");
const logger_1 = require("../utils/logger");
const manager_1 = require("./manager");
const types_1 = require("./types");
class RuntimeManagerV2 extends manager_1.BaseRuntimeManager {
    manageHealth;
    _port;
    wss;
    runtimes = new Map();
    get port() {
        return this._port;
    }
    pendingRequests = new Map();
    streamCallbacks = new Map();
    traceIdCallbacks = new Map();
    eventEmitter = new events_1.default();
    requestIdCounter = 0;
    constructor(telemetryServerUrl, manageHealth, projectRoot, processManager, disableRealtimeTelemetry = false) {
        super(telemetryServerUrl, projectRoot, processManager, disableRealtimeTelemetry);
        this.manageHealth = manageHealth;
    }
    static async create(options) {
        const manager = new RuntimeManagerV2(options.telemetryServerUrl, options.manageHealth ?? true, options.projectRoot, options.processManager, options.disableRealtimeTelemetry);
        await manager.startWebSocketServer(options.reflectionV2Port);
        return manager;
    }
    async startWebSocketServer(port) {
        if (!port) {
            port = await (0, get_port_1.default)({ port: (0, get_port_1.makeRange)(3200, 3400) });
        }
        this.wss = new ws_1.WebSocketServer({ port });
        this._port = port;
        logger_1.logger.info(`Starting reflection server: ws://localhost:${port}`);
        this.wss.on('connection', (ws) => {
            ws.on('error', (err) => logger_1.logger.error(`WebSocket error: ${err}`));
            ws.on('message', (data) => {
                try {
                    const message = JSON.parse(data.toString());
                    this.handleMessage(ws, message);
                }
                catch (error) {
                    logger_1.logger.error('Failed to parse WebSocket message:', error);
                }
            });
            ws.on('close', () => {
                this.handleDisconnect(ws);
            });
        });
        return { port };
    }
    handleMessage(ws, message) {
        if ('method' in message) {
            this.handleRequest(ws, message);
        }
        else {
            this.handleResponse(message);
        }
    }
    handleRequest(ws, request) {
        switch (request.method) {
            case 'register':
                this.handleRegister(ws, request);
                break;
            case 'streamChunk':
                this.handleStreamChunk(request);
                break;
            case 'runActionState':
                this.handleRunActionState(request);
                break;
            default:
                logger_1.logger.warn(`Unknown method: ${request.method}`);
        }
    }
    handleRegister(ws, request) {
        const params = reflection_1.ReflectionRegisterParamsSchema.parse(request.params);
        const runtimeInfo = {
            id: params.id,
            pid: params.pid,
            name: params.name,
            genkitVersion: params.genkitVersion,
            reflectionApiSpecVersion: params.reflectionApiSpecVersion,
            reflectionServerUrl: `ws://localhost:${this.port}`,
            timestamp: new Date().toISOString(),
            projectName: path_1.default.basename(this.projectRoot),
        };
        this.runtimes.set(runtimeInfo.id, { ws, info: runtimeInfo });
        this.eventEmitter.emit(types_1.RuntimeEvent.ADD, runtimeInfo);
        if (request.id) {
            ws.send(JSON.stringify({
                jsonrpc: '2.0',
                result: {
                    telemetryServerUrl: this.telemetryServerUrl,
                },
                id: request.id,
            }));
        }
    }
    handleStreamChunk(notification) {
        const { requestId, chunk } = reflection_1.ReflectionStreamChunkParamsSchema.parse(notification.params);
        const callback = this.streamCallbacks.get(requestId);
        if (callback) {
            callback(chunk);
        }
    }
    handleRunActionState(notification) {
        const { requestId, state } = reflection_1.ReflectionRunActionStateParamsSchema.parse(notification.params);
        const callback = this.traceIdCallbacks.get(requestId);
        if (callback && state?.traceId) {
            callback(state.traceId);
        }
    }
    handleResponse(response) {
        const pending = this.pendingRequests.get(response.id);
        if (pending) {
            if (response.error) {
                const errorData = response.error.data || {};
                const massagedData = {
                    ...errorData,
                    stack: errorData.details?.stack,
                    data: {
                        genkitErrorMessage: errorData.message,
                        genkitErrorDetails: errorData.details,
                    },
                };
                const error = new types_1.GenkitToolsError(response.error.message);
                error.data = massagedData;
                pending.reject(error);
            }
            else {
                let result = response.result;
                if (pending.method === 'listActions') {
                    result = reflection_1.ReflectionListActionsResponseSchema.parse(result);
                }
                else if (pending.method === 'listValues') {
                    result = reflection_1.ReflectionListValuesResponseSchema.parse(result);
                }
                else if (pending.method === 'cancelAction') {
                    result = reflection_1.ReflectionCancelActionResponseSchema.parse(result);
                }
                pending.resolve(result);
            }
            this.pendingRequests.delete(response.id);
        }
        else {
            logger_1.logger.warn(`Received response for unknown request ID ${response.id}`);
        }
    }
    handleDisconnect(ws) {
        for (const [id, runtime] of this.runtimes.entries()) {
            if (runtime.ws === ws) {
                this.runtimes.delete(id);
                this.eventEmitter.emit(types_1.RuntimeEvent.REMOVE, runtime.info);
                break;
            }
        }
    }
    async sendRequest(runtimeId, method, params) {
        const runtime = this.runtimes.get(runtimeId);
        if (!runtime) {
            throw new Error(`Runtime ${runtimeId} not found`);
        }
        switch (method) {
            case 'listValues':
                reflection_1.ReflectionListValuesParamsSchema.parse(params);
                break;
            case 'runAction':
                reflection_1.ReflectionRunActionParamsSchema.parse(params);
                break;
            case 'cancelAction':
                reflection_1.ReflectionCancelActionParamsSchema.parse(params);
                break;
        }
        const id = (++this.requestIdCounter).toString();
        const message = {
            jsonrpc: '2.0',
            method,
            params,
            id,
        };
        return new Promise((resolve, reject) => {
            const timeoutId = setTimeout(() => {
                if (this.pendingRequests.has(id)) {
                    this.pendingRequests.delete(id);
                    reject(new Error(`Request ${id} timed out`));
                }
            }, 30000);
            this.pendingRequests.set(id, {
                method,
                resolve: (value) => {
                    clearTimeout(timeoutId);
                    resolve(value);
                },
                reject: (reason) => {
                    clearTimeout(timeoutId);
                    reject(reason);
                },
            });
            runtime.ws.send(JSON.stringify(message));
        });
    }
    sendNotification(runtimeId, method, params) {
        const runtime = this.runtimes.get(runtimeId);
        if (!runtime) {
            logger_1.logger.warn(`Runtime ${runtimeId} not found, cannot send notification`);
            return;
        }
        switch (method) {
            case 'configure':
                reflection_1.ReflectionConfigureParamsSchema.parse(params);
                break;
            case 'sendInputStreamChunk':
                reflection_1.ReflectionSendInputStreamChunkParamsSchema.parse(params);
                break;
            case 'endInputStream':
                reflection_1.ReflectionEndInputStreamParamsSchema.parse(params);
                break;
        }
        const message = {
            jsonrpc: '2.0',
            method,
            params,
        };
        runtime.ws.send(JSON.stringify(message));
    }
    listRuntimes() {
        return Array.from(this.runtimes.values()).map((r) => r.info);
    }
    getRuntimeById(id) {
        return this.runtimes.get(id)?.info;
    }
    getMostRecentRuntime() {
        const runtimes = this.listRuntimes();
        if (runtimes.length === 0)
            return undefined;
        return runtimes[runtimes.length - 1];
    }
    getMostRecentDevUI() {
        return undefined;
    }
    onRuntimeEvent(listener) {
        const listeners = [];
        Object.values(types_1.RuntimeEvent).forEach((event) => {
            const fn = (rt) => listener(event, rt);
            this.eventEmitter.on(event, fn);
            listeners.push({ event, fn });
        });
        return () => {
            listeners.forEach(({ event, fn }) => {
                this.eventEmitter.off(event, fn);
            });
        };
    }
    async listActions(input) {
        const runtimeId = input?.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error(input?.runtimeId
                ? `No runtime found with ID ${input.runtimeId}.`
                : 'No runtimes found. Make sure your app is running using the `start_runtime` MCP tool or the CLI: `genkit start -- ...`. See getting started documentation.');
        }
        const result = {};
        const response = reflection_1.ReflectionListActionsResponseSchema.parse(await this.sendRequest(runtimeId, 'listActions'));
        for (const key of Object.keys(response.actions)) {
            const action = response.actions[key];
            if (!action.key) {
                action.key = key;
            }
            result[key] = action;
        }
        return result;
    }
    async listValues(input) {
        const runtimeId = input?.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error(input?.runtimeId
                ? `No runtime found with ID ${input.runtimeId}.`
                : 'No runtimes found. Make sure your app is running using `genkit start -- ...`. See getting started documentation.');
        }
        const response = await this.sendRequest(runtimeId, 'listValues', {
            type: input.type,
        });
        return response.values;
    }
    async stop() {
        if (this.wss) {
            this.wss.close();
        }
        if (this.processManager) {
            await this.processManager.kill();
        }
    }
    async runAction(input, streamingCallback, onTraceId, inputStream) {
        const runtimeId = input.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error('No runtimes found. Make sure your app is running using the `start_runtime` MCP tool or the CLI: `genkit start -- ...`. See getting started documentation.');
        }
        const runtime = this.runtimes.get(runtimeId);
        if (!runtime) {
            throw new Error(`Runtime ${runtimeId} not found`);
        }
        const id = (++this.requestIdCounter).toString();
        if (streamingCallback) {
            this.streamCallbacks.set(id, streamingCallback);
        }
        if (onTraceId) {
            this.traceIdCallbacks.set(id, onTraceId);
        }
        const message = {
            jsonrpc: '2.0',
            method: 'runAction',
            params: {
                ...input,
                stream: !!streamingCallback,
                streamInput: !!inputStream,
            },
            id,
        };
        const promise = new Promise((resolve, reject) => {
            this.pendingRequests.set(id, { method: 'runAction', resolve, reject });
            runtime.ws.send(JSON.stringify(message));
        })
            .then((result) => {
            return action_1.RunActionResponseSchema.parse(result);
        })
            .finally(() => {
            if (streamingCallback) {
                this.streamCallbacks.delete(id);
            }
            if (onTraceId) {
                this.traceIdCallbacks.delete(id);
            }
        });
        if (inputStream) {
            (async () => {
                try {
                    for await (const chunk of inputStream) {
                        this.sendNotification(runtimeId, 'sendInputStreamChunk', {
                            requestId: id,
                            chunk,
                        });
                    }
                    this.sendNotification(runtimeId, 'endInputStream', {
                        requestId: id,
                    });
                }
                catch (e) {
                    logger_1.logger.error(`Error streaming input: ${e}`);
                }
            })();
        }
        return promise;
    }
    async cancelAction(input) {
        const runtimeId = input.runtimeId || this.getMostRecentRuntime()?.id;
        if (!runtimeId) {
            throw new Error('No runtime found');
        }
        return this.sendRequest(runtimeId, 'cancelAction', {
            traceId: input.traceId,
        });
    }
}
exports.RuntimeManagerV2 = RuntimeManagerV2;
